src/routes/api/user/index.ts

```diff
@@ -0,0 +1,60 @@
+import {
+  FastifyPluginAsyncTypebox,
+  Type
+} from '@fastify/type-provider-typebox'
+import { Auth } from '../../../schemas/auth.js'
+import { UpdateCredentialsSchema } from '../../../schemas/users.js'
+
+const plugin: FastifyPluginAsyncTypebox = async (fastify) => {
+  fastify.put(
+    '/update-password',
+    {
+      schema: {
+        body: UpdateCredentialsSchema,
+        response: {
+          200: Type.Object({
+            message: Type.String()
+          }),
+          401: Type.Object({
+            message: Type.String()
+          })
+        },
+        tags: ['User']
+      }
+    },
+    async function (request, reply) {
+      const { newPassword, currentPassword } = request.body
+      const username = request.session.user.username
+
+      try {
+        const user = await fastify.knex<Auth>('users')
+          .select('username', 'password')
+          .where({ username })
+          .first()
+
+        if (user) {
+          const isPasswordValid = await fastify.compare(currentPassword, user.password)
+
+          if (isPasswordValid) {
+            const hashedPassword = await fastify.hash(newPassword)
+
+            await fastify.knex('users')
+              .update({
+                password: hashedPassword
+              })
+              .where({ username })
+
+            return { message: 'Password updated successfully' }
+          }
+        }
+
+        reply.status(401)
+        return { message: 'Invalid username or password.' }
```

Comment:
This error message is wrong.

Reply:
> This error message is wrong.
I think we should not return to the client if the user or password is wrong, it is not a good security design to prevent potential attackers from getting information about the part of the login information that is incorrect.

Or i have missed something

Reply:
> I think we should not return to the client if the user or password is wrong

For a auth endpoint yes, for a setting endpoint, yes, because we know this is the user/admin that do this operation (unless hacked, this is why we have rate limiter). We must inform why the request for password update failed.

Regarding the line `52`, if there is no user session, the error should be detected ahead of time in this hook:
https://github.com/fastify/demo/blob/4470023655ea4e68de65d9b4e78a6f03c56ab49a/src/routes/api/autohooks.ts#L9

But you can still return: ` reply.unauthorized()`, it will set status code to `401` automatically.